lib/ProductOpener/Tags.pm

```diff
@@ -807,33 +897,73 @@ sub build_tags_taxonomy($$$) {
+	# it's the tag value where we lowercased, replace separators by dash etc.,
+	# see get_string_id_for_lang
+
+	# line tagid (or tagid_c) in particular is the first entry in a line
+	# that is the id for this tag in a particular language
+
+	# when we speak about canonical tagid (or canon_tagid),
+	# this is the line tagid for the first line of a definition block,
+	# This is the id for the tag among languages,
+	# we keep the language code as prefix, as it may be in any language
 
 	# Need to be initialized as a taxonomy is probably already loaded by Tags.pm
+	# stopwords contains three entry per language
+	# * <lc> contains an array of normalized stop words
+	# * <lc>.strings contains an array of stop words
+	# * <lc>.orig the original lines as a string
 	$stopwords{$tagtype} = {};
+	# synonyms track know synonyms and associate to their tagid, by language
+	# Note that it contains synonyms and extended synonyms
+	# tagtype -> lc -> tagid stores the line tagid
+	# Note: it could have been named synonym_of
 	$synonyms{$tagtype} = {};
+	# synonyms by language for each tagid (this is the reverse lookup of synonyms)
+	# but only for direct synonyms (not extended one)
+	# tagtype -> lc -> line_tagid stores synonyms tag strings
 	$synonyms_for{$tagtype} = {};
+	# this is a close parent to synonyms_for,
+	# but this contains synonyms generated by substitutions
+	# tagtype -> lc -> line_tagid -> tagid contains 1 if tagid in a synonym of line_tagid
 	$synonyms_for_extended{$tagtype} = {};
+	# given a tagid, with language prefix, gives the corresponding canonical tag id
+	# tagtype -> lc:tagid gives you canon_tagid
 	$translations_from{$tagtype} = {};
+	# associtate each canonical tag id, to a hash where key is language code and value the tag text
+	# eg: tagtype -> canon_tagid -> lc gives you text tag
 	$translations_to{$tagtype} = {};
 	$level{$tagtype} = {};
+	# list of parents for every tag as a hashmap (only canonical tagid)
+	# $tagtype -> $canon_tagid -> $parentid contains 1
 	$direct_parents{$tagtype} = {};
+	# list of children for every tag as a hashmap
+	# $tagtype -> $canon_tagid -> $childid contains 1
 	$direct_children{$tagtype} = {};
 	$all_parents{$tagtype} = {};
 	$root_entries{$tagtype} = {};
-
+	# a list of all canon_tagid as a hashmap
+	# $tagtype -> $canon_tagid contains 1 for every canonical tagid
 	$just_tags{$tagtype} = {};
 	# synonyms that are not real entries, but only enrich existing tags
 	# they correspond to synonyms: entries
+	# this is a hashmap where keys are canonical tagid, and value is 1
 	$just_synonyms{$tagtype} = {};
+	# this stores properties for each canonical tagid
+	# $tagtype -> $canon_tagid -> "$property:$lc" stores the value for property
 	$properties{$tagtype} = {};
 
 	my $errors = '';
 
 	if (open (my $IN, "<:encoding(UTF-8)", "$data_root/taxonomies/$file")) {
 
-		my $current_tagid;
-		my $current_tag;
-		my $canon_tagid;
+		my $line_tagid;  # tracks the tag normalized form
```

Comment:
$line_tagid is the normalized form of $current_tag, so they should have the same prefix.

Maybe we can be more specific:

my $lc_tag; # Main name of a tag in a specific language (display form) - e.g. "Café au lait"
my $lc_tagid; # Main id of a tag in a specific language (normalized form) - e.g. "cafe-au-lait"
my $canon_tagid; # Canonical id of the tag (main language prefix + normalized form in the main language) - e.g. "en:coffee-with-milk"

Reply:
I would not use "line" in the variable name.

Reply:
ok, I think clarity is important.

Reply:
done (next commit)